torch/nativert/graph/Graph.h

```diff
@@ -0,0 +1,696 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include <ATen/core/ivalue.h>
+#include <c10/util/IntrusiveList.h>
+#include <c10/util/Logging.h>
+
+#include <torch/csrc/utils/generated_serialization_types.h>
+#include <torch/nativert/executor/Placement.h>
+#include <torch/nativert/graph/GraphSignature.h>
+#include <torch/nativert/graph/TensorMeta.h>
+
+namespace torch::nativert {
+
+using NodeIndex = size_t;
+
+class Value;
+
+class Type {
+ public:
+  enum Kind {
+    None,
+    Tensor,
+    TensorList,
+    OptionalTensorList,
+    SymInt,
+    SymIntList,
+    SymBool,
+    SymFloat,
+    CustomObj,
+  };
+
+  friend std::ostream& operator<<(std::ostream& out, const Type& ty);
+
+  /* implicit */ Type(Kind kind) : kind_(kind) {}
+
+  explicit Type(Kind kind, const std::string& classFqn)
+      : kind_(kind), classFqn_(classFqn) {
+    CHECK(kind == Kind::CustomObj);
+    CHECK(!classFqn.empty());
+  }
+
+  Kind kind() const {
+    return kind_;
+  }
+
+  std::string classFqn() const {
+    CHECK(kind_ == Kind::CustomObj)
+        << "Only CustomObj type can have classFqn, got " << kind_;
+    return classFqn_;
+  }
+
+ private:
+  const Kind kind_;
+  const std::string classFqn_;
+};
+
+bool operator==(const Type& left, const Type& right);
+
+// These are all the constant types that are allowed as attributes on Nodes.
+struct None {};
+// None always equals itself
+inline bool operator==(const None&, const None&) {
+  return true;
+}
+
+class Graph;
+
+using Constant = std::variant<
+    None,
+    int64_t,
+    std::vector<int64_t>,
+    double,
+    std::vector<double>,
+    std::string,
+    c10::ScalarType,
+    c10::MemoryFormat,
+    c10::Layout,
+    c10::Device,
+    bool,
+    std::vector<bool>,
+    std::vector<std::string>,
+    std::unique_ptr<Graph>>;
+
+c10::IValue constantToIValue(const Constant& constant);
+
+class Node;
+
+/**
+ * Represents a single symbolic value (tensor/symint/list of them). Values are
+ * inputs and outputs of Nodes.
+ */
+using ValueId = int;
+class Value {
+ public:
+  explicit Value(ValueId id, std::string name, Type t, Node* producer)
+      : name_(std::move(name)), id_(id), type_(t), producer_(producer) {
+    TORCH_CHECK_EQ(name_, this->name());
```

Comment:
why do we need this check? name() returns name_. if there is some debugging reason we need it, how about TORCH_INTERNAL_ASSERT_DEBUG_ONLY?